flac.go

```diff
@@ -62,73 +68,105 @@ type Stream struct {
 // Stream.ParseNext to parse the entire next frame including audio samples.
 func New(r io.Reader) (stream *Stream, err error) {
 	// Verify FLAC signature and parse the StreamInfo metadata block.
-	br := bufio.NewReader(r)
-	stream = &Stream{r: br}
-	isLast, err := stream.parseStreamInfo()
+	stream = &Stream{}
+
+	rs, seeker := r.(io.ReadSeeker)
+	if seeker {
+		stream.r = r
+	} else {
+		stream.r = bufio.NewReader(r)
```

Comment:
What do you think about this change?

Here is the reasoning:
1.  if stream.r  is *bufio.Reader then it is no longer possible to assert for io.ReadSeeker and so Seek is no longer possible.
2. if r is io.ReadSeeker then r must be either *os.File or an in-memory buffer of some kind.
3. if r is not io.ReadSeeker then it is probably a network stream, like from http.Request.Body???

My guess is that bufio.Reader is really only needed when the reader streaming over a network.  If reader is *os.File or it's an in memory buffer then bufio.Reader really isn't necessary.

Reply:
> What do you think about this change?

The change looks good to me, and the reasoning valid.

> My guess is that bufio.Reader is really only needed when the reader is streaming over a network. If reader is `*os.File` or it's an in memory buffer then bufio.Reader really isn't necessary.

I think we did some preliminary tests using `*os.File` without buffered reading and it did (at least for an older version of `mewkiz/flac`) impact the performance. This was the primary reason we used buffered reading as default.

Reply:
If *os.File did cause problems then it seems like explicitly opting in for the ability to Seek is the thing to do.

something like?

`type Option func(stream *Stream) (err error)

func New(r io.Reader, opts ...Option) (stream *Stream, err error) {
   ...
}

func EnableSeek(stream *Stream) (err error) {
   // don't make stream.r a *bufio.Reader and look for seekTable in metadata
   ...
}
`